scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,622 @@
+///
+/// Asynchronous code is rarely useful if it is serialized. In practice,
+/// you will want to spawn asynchronous tasks or in the case of Scipio which
+/// support multi-queued execution have a dedicated [`TaskQueue`] for each group
+/// of tasks with their own priority.
+///
+/// The `LocalChannel` is an abstraction similar to the [`channel`] that can be
+/// used for this purpose
+///
+/// The communication between sender and receiver is broken when one of them goes
+/// out of scope. They however behave differently:
+///
+/// * The [`LocalReceiver`] never sees an error, as it is implemented as a stream
+///   interface compatible with [`StreamExt`]. When the sender is no longer available the receiver's call to [`next`] will return [`None`].
+/// * The [`LocalSender`] will return a [`ChannelError`] encapsulating a [`BrokenPipe`] error if it tries to [`push`] into
+///   a channel that no longer has a receiver.
+///
+/// # Examples
+///
+/// ```
+/// use scipio::{LocalExecutor, Local, Latency, Shares};
+/// use scipio::channels::LocalChannel;
+/// use futures::stream::StreamExt;
+///
+/// let ex = LocalExecutor::make_default();
+/// ex.run(async move {
+///     let task_queue =
+///         Local::create_task_queue(
+///             Shares::default(), Latency::NotImportant, "example");
+///     
+///     let (sender, mut receiver) = LocalChannel::new_unbounded();
+///     let h = Local::local_into(async move {
+///         assert_eq!(receiver.next().await.unwrap(), 0);
+///     }, task_queue).unwrap().detach();
+///     sender.try_send(0);
+///     drop(sender);
+///     h.await;
+/// });
+/// ```
+/// [`TaskQueue`]: struct.TaskQueue.html
+/// [`LocalSender`]: struct.LocalSender.html
+/// [`LocalReceiver`]: struct.LocalReceiver.html
+/// [`ChannelError`]: struct.ChannelError.html
+/// [`push`]: struct.LocalSender.html#method.push
+/// [`None`]: https://doc.rust-lang.org/std/option/enum.Option.html#variant.None
+/// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+/// [`channel`]: https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html
+/// [`next`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.next
+/// [`StreamExt`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html
+pub struct LocalChannel<T> {
+    state: Rc<RefCell<State<T>>>,
+}
+
+impl<T> LocalChannel<T> {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(capacity: ChannelCapacity) -> (LocalSender<T>, LocalReceiver<T>) {
+        let channel = match capacity {
+            ChannelCapacity::Unbounded => VecDeque::new(),
+            ChannelCapacity::Bounded(x) => VecDeque::with_capacity(x),
+        };
+
+        let state = Rc::new(RefCell::new(State {
+            capacity,
+            channel,
+            send_waiters: Vec::new(),
+            recv_waiters: Vec::new(),
+            receiver_alive: true,
+            sender_alive: true,
```

Comment:
I wonder if it makes sense to pair `waiters: Vec` and `alive: bool` into a single `waiters: Option<Vec>`?

That way, type system guarantees that you can‘t add waiter if the opposite side is closed.